pyro/ops/welford.py

```diff
@@ -0,0 +1,70 @@
+from collections import OrderedDict
+
+import torch
+
+
+class WelfordCovariance(object):
+    """
+    Implements Welford's online scheme for estimating (co)variance (see :math:`[1]`).
+    Useful for adapting diagonal and dense mass structures for HMC.
+
+    **References**
+
+    [1] `The Art of Computer Programming`,
+    Donald E. Knuth
+    """
+    def __init__(self, diagonal=True):
+        self.diagonal = diagonal
+        self.reset()
+
+    def _deltas(self, z):
+        deltas = OrderedDict()
```

Comment:
Since we are using `OrderedDict` for `z` and `r` in HMC and NUTS, can we simplify the interface of `WelfordCovariance` to use lists instead, and just pass in `z.values()` here? It will be nice to keep utilities in `ops` independent of constructs like `site_name` which are still specific to Pyro inference.